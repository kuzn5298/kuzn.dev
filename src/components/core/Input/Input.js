import React from 'react';
import clsx from 'clsx';

import { InputContainer } from './Input.styled';

const Input = ({ errorMessage, error, className, ...inputProps }, ref) => (
    <InputContainer className={className}>
        <input className={clsx('input', error && 'error-container')} ref={ref} {...inputProps} />
        {error && errorMessage && <div className="error">{errorMessage}</div>}
    </InputContainer>
);

export default React.forwardRef(Input);
