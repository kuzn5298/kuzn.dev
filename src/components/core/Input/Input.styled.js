import styled from 'styled-components';

export const InputContainer = styled.div`
    display: inline-block;
    width: ${({ fullWidth }) => (fullWidth ? '100%' : 'auto')};
    overflow: hidden;

    .input {
        background-color: transparent;
        outline: 0;
        border: 0;
        margin: 0;
        padding: 1rem 1.5rem;
        background-color: ${({ theme }) => theme.palette.common.gray};
        box-sizing: border-box;
        border: 2px solid ${({ theme }) => theme.palette.common.gray};
        width: 100%;

        &.error-container {
            border-color: ${({ theme }) => theme.palette.common.error};
        }

        &:focus {
            border-color: ${({ theme }) => theme.palette.primary.main};
        }
    }

    .error {
        color: ${({ theme }) => theme.palette.common.error};
    }
`;
