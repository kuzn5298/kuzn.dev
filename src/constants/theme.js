import { MoonIcon, SunIcon } from '@components/icons';

export const THEME = {
    LIGHT: 'light',
    DARK: 'dark',
};

export const NEXT_THEME = {
    [THEME.DARK]: {
        id: THEME.LIGHT,
        name: 'Light',
        icon: SunIcon,
    },
    [THEME.LIGHT]: {
        id: THEME.DARK,
        name: 'Dark',
        icon: MoonIcon,
    },
};
