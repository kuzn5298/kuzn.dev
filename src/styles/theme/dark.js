import _ from 'lodash';
import defaultTheme from './default';

const darkTheme = {
    palette: {
        background: {
            default: '#1D1C1C',
        },
        text: {
            primary: '#F2F2F2',
        },
        primary: {
            main: '#B3E5FC',
        },
    },
};

export default _.merge({}, defaultTheme, darkTheme);
