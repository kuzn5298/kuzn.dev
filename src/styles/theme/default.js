const defaultTheme = {
    typography: {
        fontFamily: 'Lexend, Arial, Helvetica, sans-serif;',
        fontSize: 16,
    },
};

export default defaultTheme;
