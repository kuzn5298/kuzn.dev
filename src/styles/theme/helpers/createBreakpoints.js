const STEP = 5 / 100;

const createBreakpoints = (values, unit = 'px') => {
    const keys = Object.keys(values);

    const up = (key) => {
        const value = values[key] ?? key;
        return `@media (min-width: ${value}${unit})`;
    };

    const down = (key) => {
        const value = values[key] ?? key;
        return `@media (max-width: ${value - STEP}${unit})`;
    };

    return {
        keys,
        values,
        up,
        down,
        unit,
    };
};

export default createBreakpoints;
