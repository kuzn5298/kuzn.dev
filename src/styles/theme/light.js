import _ from 'lodash';
import defaultTheme from './default';

const lightTheme = {
    palette: {
        background: {
            default: '#fff',
        },
        text: {
            primary: '#000',
        },
        primary: {
            main: '#B3E5FC',
        },
    },
};

export default _.merge({}, defaultTheme, lightTheme);
