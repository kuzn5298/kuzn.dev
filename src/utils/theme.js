import { THEME, DEFAULT_THEME_ID, THEME_LOCAL_STORAGE, IS_BROWSER } from '@constants';
import storage from '@libs/storage';

export const isValidThemeId = (themeId) => Object.values(THEME).includes(themeId);

export const setThemeIdToLocalStorages = (themeId) => {
    storage.set(THEME_LOCAL_STORAGE, themeId);
};

export const getThemeIdFromLocalStorages = () => {
    const themeId = storage.get(THEME_LOCAL_STORAGE);
    return isValidThemeId(themeId) ? themeId : null;
};

export const getSystemTheme = () => {
    const isDarkTheme = IS_BROWSER
        ? window.matchMedia('(prefers-color-scheme: dark)')?.matches
        : false;
    return isDarkTheme ? THEME.DARK : null;
};

export const getUsedThemeId = () => {
    const localTheme = getThemeIdFromLocalStorages();
    const systemTheme = getSystemTheme();
    return localTheme ?? systemTheme ?? DEFAULT_THEME_ID;
};
