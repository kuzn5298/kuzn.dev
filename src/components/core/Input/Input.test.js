import React from 'react';
import { wrapWithTheme } from '@tests/helpers';
import { lightTheme } from '@styles/theme';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Input from './Input';

describe('<Input />', () => {
    it('should render component', () => {
        const { getByRole } = render(wrapWithTheme(<Input />));
        expect(getByRole('textbox')).toBeInTheDocument();
    });

    it('should have default class', () => {
        const { getByRole } = render(wrapWithTheme(<Input />));
        expect(getByRole('textbox')).toHaveClass('input');
    });

    it('should display error message when error prop is true', () => {
        const errorMessage = 'This is an error';
        const { getByText } = render(wrapWithTheme(<Input error errorMessage={errorMessage} />));

        const inputElement = getByText(errorMessage);

        expect(inputElement).toBeInTheDocument();
        expect(inputElement).toHaveClass('error');
    });

    it('should not display error message when error prop is false', () => {
        const errorMessage = 'This is an error';
        const { queryByText } = render(wrapWithTheme(<Input errorMessage={errorMessage} />));
        expect(queryByText(errorMessage)).not.toBeInTheDocument();
    });

    it('should apply error class when error prop is true', () => {
        const { getByRole } = render(wrapWithTheme(<Input error />));
        expect(getByRole('textbox')).toHaveClass('error-container');
    });

    it('should render an input with different props (snapshot test)', () => {
        const errorMessage = 'This is an error';

        const { asFragment } = render(
            wrapWithTheme(<Input error placeholder="Snapshot test" errorMessage={errorMessage} />)
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('should focus input on tab press and change border color on focus (event)', async () => {
        const { getByRole } = render(wrapWithTheme(<Input />, lightTheme));
        const inputElement = getByRole('textbox');

        expect(inputElement).not.toHaveFocus();
        await userEvent.tab();
        expect(inputElement).toHaveFocus();
        expect(inputElement).toHaveStyle(`border-color: ${lightTheme.palette.primary.main}`);
    });

    it('should allow typing into input field (event)', async () => {
        const text = 'Hello world';
        const { getByRole } = render(wrapWithTheme(<Input />));
        const inputElement = getByRole('textbox');

        await userEvent.type(inputElement, text);
        expect(inputElement).toHaveValue(text);
    });

    it('should call onChange on typing (event)', async () => {
        const handleChange = jest.fn();
        const text = 'Hello world';
        const { getByRole } = render(wrapWithTheme(<Input onChange={handleChange} />));
        const inputElement = getByRole('textbox');

        await userEvent.type(inputElement, text);
        expect(handleChange).toHaveBeenCalledTimes(text.length);
    });

    it('should call onFocus when input is focused (event)', async () => {
        const handleFocus = jest.fn();
        const { getByRole } = render(wrapWithTheme(<Input onFocus={handleFocus} />));
        const inputElement = getByRole('textbox');

        await userEvent.click(inputElement);

        expect(handleFocus).toHaveBeenCalled();
    });

    it('should call onBlur when input loses focus (event)', async () => {
        const handleBlur = jest.fn();
        const { getByRole } = render(wrapWithTheme(<Input onBlur={handleBlur} />));

        const inputElement = getByRole('textbox');

        await userEvent.click(inputElement);
        await userEvent.tab();

        expect(handleBlur).toHaveBeenCalled();
    });

    test('should call onKeyUp and onKeyDown when a key is released (event)', async () => {
        const handleKeyUp = jest.fn();
        const handleKeyDown = jest.fn();
        const { getByRole } = render(
            wrapWithTheme(<Input onKeyUp={handleKeyUp} onKeyDown={handleKeyDown} />)
        );

        const inputElement = getByRole('textbox');

        await userEvent.type(inputElement, 'A');

        expect(handleKeyUp).toHaveBeenCalled();
        expect(handleKeyDown).toHaveBeenCalled();
    });
});
