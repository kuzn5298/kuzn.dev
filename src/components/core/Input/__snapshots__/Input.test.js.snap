// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should render an input with different props (snapshot test) 1`] = `
<DocumentFragment>
  .c0 {
  display: inline-block;
  overflow: hidden;
}

.c0 .input {
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  padding: 1rem 1.5rem;
  background-color: #cdcdcd;
  box-sizing: border-box;
  border: 2px solid #cdcdcd;
  width: 100%;
}

.c0 .input.error-container {
  border-color: #e53935;
}

.c0 .input:focus {
  border-color: #03a9f4;
}

.c0 .error {
  color: #e53935;
}

<div
    class="c0"
  >
    <input
      class="input error-container"
      placeholder="Snapshot test"
    />
    <div
      class="error"
    >
      This is an error
    </div>
  </div>
</DocumentFragment>
`;
